File: ports/go/adapters/langchaingo/hitl.go
package langchaingo

import "context"

// InterruptError is the HITL pause signal (PROTOCOL.md §5) — return it from a
// Tool.Execute (via Interrupt) and the runtime stores the pending call, emits
// a botiva interrupt event (approval chips in the client) and ends the turn.
// Not an error in the failure sense.
type InterruptError struct {
	Payload any // recommended: map with {question, options} → rendered as chips
}

func (e *InterruptError) Error() string { return "botiva interrupt (waiting for user input)" }

// resumeHolder carries the user's answer into the re-run of the paused tool.
// One-shot: a second Interrupt in the same run pauses again.
type resumeHolder struct {
	value string
	used  bool
}

type resumeKey struct{}

func withResume(ctx context.Context, value string) context.Context {
	return context.WithValue(ctx, resumeKey{}, &resumeHolder{value: value})
}

// Interrupt is the Go counterpart of LangGraph's interrupt() for this adapter.
//
// First pass: returns an *InterruptError — propagate it as the tool error and
// the run pauses. Resume pass (the user's next message): returns the answer.
//
//	func generateReport(ctx context.Context, args map[string]any) (string, error) {
//	    answer, err := langchaingo.Interrupt(ctx, map[string]any{
//	        "question": "Generate the report?", "options": []string{"Approve", "Cancel"},
//	    })
//	    if err != nil { return "", err } // pause
//	    if !approveRe.MatchString(answer) { return "The user declined.", nil }
//	    ...
//	}
func Interrupt(ctx context.Context, payload any) (string, error) {
	if holder, ok := ctx.Value(resumeKey{}).(*resumeHolder); ok && !holder.used {
		holder.used = true
		return holder.value, nil
	}
	return "", &InterruptError{Payload: payload}
}

File: packages/runtimes/dotnet/Hitl.cs
namespace Botiva.Agents;

/// <summary>
/// Thrown by <see cref="Hitl.Interrupt"/> on the first pass through a tool —
/// the runtime turns it into a botiva interrupt event (approval chips in the
/// client) and pauses the run. Not an error.
/// </summary>
public sealed class BotivaInterruptException(object? payload) : Exception("botiva interrupt (waiting for user input)")
{
    public object? Payload { get; } = payload;
}

/// <summary>
/// LangGraph-style interrupt() for .NET tools (PROTOCOL.md §5).
///
/// First pass: throws <see cref="BotivaInterruptException"/> — the engine
/// stores the pending interrupt and the client shows chips. The user's next
/// message re-runs the SAME tool call with the answer injected, so the second
/// pass returns it:
///
///   [Description("Generates a report PDF (asks the user for approval first).")]
///   static string GenerateReportPdf(string topic)
///   {
///       var answer = (string?)Hitl.Interrupt(new { question = $"Generate \"{topic}\"?", options = new[] { "Approve", "Cancel" } });
///       if (!Regex.IsMatch(answer ?? "", "approve|yes|onay|evet", RegexOptions.IgnoreCase))
///           return "The user declined — no report was generated.";
///       return $"Report ready: report-{topic}.pdf";
///   }
/// </summary>
public static class Hitl
{
    private static readonly AsyncLocal<object?> Resume = new();

    /// <summary>
    /// Pause for user input. Returns the user's answer on the resume pass;
    /// throws <see cref="BotivaInterruptException"/> on the first pass.
    /// Recommended payload: <c>{ question, options }</c> → rendered as chips.
    /// </summary>
    public static object? Interrupt(object? payload)
    {
        if (Resume.Value is { } answer)
        {
            Resume.Value = null; // one-shot: a second Interrupt in the same run pauses again
            return answer;
        }
        throw new BotivaInterruptException(payload);
    }

    /// <summary>Runtime hook: make <paramref name="value"/> available to the next Interrupt call downstream.</summary>
    internal static void SetResume(object? value) => Resume.Value = value;
}
